File: Python/10-BitManipulation/Algorithms/3-Checking-if-a-number-is-a-power-of-2/Checking-if-a-number-is-a-power-of-2Pro.py
def is_power_of_two_approach1(n):
    """
    Approach 1: Using the property that powers of 2 have only one bit set.

    Real-world application:
    - Optimizing resource allocation: In systems where resources are allocated in powers of 2 (e.g., memory pages, buffer sizes),
      this method can quickly validate if a requested size is a valid power of 2, preventing fragmentation and ensuring efficient use.
    - Implementing binary trees or heaps: Checking for power of 2 is crucial in data structures like complete binary trees or heaps,
      where the number of nodes at each level is a power of 2.

    Time Complexity: O(1) - Bitwise operations are constant time.
    Space Complexity: O(1) - No extra space is used.
    """
    if n <= 0:  # Handle non-positive numbers as they are not powers of 2
        return False
    return (n & (n - 1)) == 0  # Key idea: (n & (n-1)) clears the least significant bit.
    #  For powers of 2, this results in 0.

def is_power_of_two_approach2(n):
    """
    Approach 2: Repeatedly dividing by 2.

    Real-world application:
    - Image processing: Some image formats or algorithms might require dimensions (width, height) to be powers of 2.
      This method can validate image dimensions before processing or resizing.
    - Audio processing:  Buffer sizes in audio processing are often powers of 2 for efficient FFT (Fast Fourier Transform) calculations.
      This method can validate buffer sizes.

    Time Complexity: O(log n) - Number of divisions depends on the magnitude of n.
    Space Complexity: O(1) - No extra space is used.
    """
    if n <= 0:
        return False
    while n % 2 == 0:
        n //= 2  # Integer division to handle Python's behavior with large numbers
    return n == 1

def is_power_of_two_approach3(n):
    """
    Approach 3: Using logarithms.

    Real-world application:
     - Scientific computing: In simulations or modeling, you might need to check if a certain parameter or grid size
       is a power of 2 for computational efficiency.
     - Data binning or bucketing: When creating histograms or binning data, you might want the number of bins to be a power of 2
       for balanced distribution.

    Time Complexity: O(1) - Logarithm calculation is generally considered constant time.
    Space Complexity: O(1) - No extra space is used.
    """
    import math  # Import the math module

    if n <= 0:
        return False
    # Calculate log base 2.  Use math.log2 for better precision
    log_result = math.log2(n)
    # Check if the result is very close to an integer
    return abs(log_result - round(log_result)) < 1e-10 # Use a small tolerance for floating-point comparison

def is_power_of_two_approach4(n):
    """
    Approach 4:  Bit manipulation with left shift.

    Real-world application:
    - Embedded systems: In low-level programming, especially in embedded systems, generating powers of 2
      using bit shifts is common for configuring hardware registers or memory addresses. This function could validate
      if a given value matches a pre-calculated power of 2.
    - Network protocols: Some network protocols use fields with lengths that are powers of 2.  This can validate field lengths.

    Time Complexity: O(log n) - The loop iterates a maximum of log2(n) times.
    Space Complexity: O(1) - No extra space is used.
    """
    if n <= 0:
        return False
    power = 1
    while power < n:
        power <<= 1  # Left shift (multiply by 2)
    return power == n

def is_power_of_two_approach5(n):
    """
    Approach 5: Using regular expressions (string representation).

    Real-world application:
    - Input validation: When reading configuration files or user input, you might receive numbers as strings.
      This method can validate if the string represents a power of 2.  For example, validating input in a server configuration.
    - Data serialization/deserialization:  When parsing data from a file or network stream, you might need to check if a numerical
      value represented as a string is a power of 2.

    Time Complexity: O(log n) - String conversion is O(log n), regex is then applied
    Space Complexity: O(log n) -  String conversion creates a string of length log n.
    """
    import re

    if n <= 0:
        return False
    binary_representation = bin(n)[2:]  # Convert to binary string, remove "0b" prefix
    # Regular expression to check for "1" followed by all "0"s
    return re.match(r"^10*$", binary_representation) is not None

def is_power_of_two_approach6(n):
    """
    Approach 6: Using bit manipulation and counting set bits.

    Real-world application:
    - Error detection/correction: In some error detection/correction codes, the number of set bits in a data word
      can be used to detect errors.  This approach can be adapted to count set bits and check if only one bit is set.
    - Cryptography:  Bit manipulation is fundamental in cryptography.  Checking the number of set bits can be a component
      in some cryptographic algorithms.

    Time Complexity: O(log n) in the worst case, where n has many set bits.  Can be O(1) in some cases.
    Space Complexity: O(1) - No extra space is used.
    """
    if n <= 0:
        return False
    count = 0
    while n > 0:
        n &= (n - 1)  # Clear the least significant bit
        count += 1
    return count == 1 # A power of two has only one bit set

def main():
    """
    Main function to demonstrate all approaches with various test cases.
    """
    test_cases = [0, 1, 2, 3, 4, 8, 15, 16, 32, 64, 100, 128, 256, -1, -2, 1024, 2048, 4096]

    print("Checking if numbers are powers of 2:")
    print("-" * 50)

    for num in test_cases:
        print(f"\nNumber: {num}")
        print(f"Approach 1: {is_power_of_two_approach1(num)}")
        print(f"Approach 2: {is_power_of_two_approach2(num)}")
        print(f"Approach 3: {is_power_of_two_approach3(num)}")
        print(f"Approach 4: {is_power_of_two_approach4(num)}")
        print(f"Approach 5: {is_power_of_two_approach5(num)}")
        print(f"Approach 6: {is_power_of_two_approach6(num)}")
        print("-" * 50)

if __name__ == "__main__":
    main()

File: C++/10-BitManipulation/Algorithms/3-Checking-if-a-number-is-a-power-of-2/Checking-if-a-number-is-a-power-of-2Pro.cpp
#include <iostream>
#include <cmath>
#include <bitset> // Required for approach 5

using namespace std;

// Function to check if a number is a power of 2 using the subtraction and bitwise AND method.
// This is generally the most efficient and widely used method.
bool isPowerOfTwo_Approach1(unsigned int n) {
    if (n == 0) return false; // 0 is not a power of 2
    return (n & (n - 1)) == 0;
    // Explanation:
    // A power of 2 in binary has only one bit set (e.g., 8 = 1000, 16 = 10000).
    // Subtracting 1 from a power of 2 will result in a number with all bits set up to the original '1' bit (e.g., 8-1 = 7 = 0111, 16-1 = 15 = 01111).
    // The bitwise AND of these two numbers will be 0.
    // For example:
    //  n = 8 (1000 in binary)
    //  n - 1 = 7 (0111 in binary)
    //  n & (n - 1) = 1000 & 0111 = 0000 = 0
    // If n is not a power of 2, the result will not be 0.
    // For example:
    //  n = 6 (0110 in binary)
    //  n - 1 = 5 (0101 in binary)
    //  n & (n - 1) = 0110 & 0101 = 0100 != 0
}

// Function to check if a number is a power of 2 by counting the set bits.
// This approach iterates through the bits of the number.
bool isPowerOfTwo_Approach2(unsigned int n) {
    if (n == 0) return false;
    int count = 0;
    while (n > 0) {
        count += (n & 1); // Check the least significant bit
        n >>= 1;             // Right shift n by 1
    }
    return count == 1;
    // Explanation:
    // A power of 2 has only one bit set to 1. This method counts the number of set bits.
    // If the count is exactly 1, then the number is a power of 2.
}

// Function to check if a number is a power of 2 using the log2 function.
// This approach uses floating-point operations, which can sometimes have precision issues.
bool isPowerOfTwo_Approach3(unsigned int n) {
    if (n == 0) return false;
    double logValue = log2(n);
    return (ceil(logValue) == floor(logValue));
    // Explanation:
    // The base-2 logarithm of a power of 2 is an integer (e.g., log2(8) = 3, log2(16) = 4).
    // We calculate the log base 2 of the number. If the ceiling and floor of the result are equal,
    // then the number is a power of 2.  We use ceil and floor to handle potential
    // floating-point inaccuracies.
}

// Function to check if a number is a power of 2 using repeated division.
// This approach is straightforward but can be less efficient for large numbers.
bool isPowerOfTwo_Approach4(unsigned int n) {
    if (n == 0) return false;
    if (n == 1) return true; // 1 is a power of 2 (2^0)
    while (n % 2 == 0) {
        n /= 2;
    }
    return n == 1;
    // Explanation:
    // We repeatedly divide the number by 2. If the number is a power of 2,
    // it will eventually become 1 after repeated division. If at any point the
    // remainder is not 0, or the final result is not 1, it's not a power of 2.
}

// Function to check if a number is a power of 2 using bitset.
bool isPowerOfTwo_Approach5(unsigned int n) {
    if (n == 0) return false;
    bitset<32> b(n); // Assuming 32-bit unsigned int
    return b.count() == 1;
    // Explanation:
    //  We convert the unsigned integer to a bitset.  The bitset's count() method
    //  returns the number of set bits.  A power of two has only one set bit.
}

int main() {
    unsigned int number;

    cout << "Enter a non-negative integer: ";
    cin >> number;

    cout << number << " is ";
    if (isPowerOfTwo_Approach1(number)) {
        cout << "a power of 2 (Approach 1: Bitwise AND)" << endl;
    } else {
        cout << "not a power of 2 (Approach 1: Bitwise AND)" << endl;
    }

    cout << number << " is ";
    if (isPowerOfTwo_Approach2(number)) {
        cout << "a power of 2 (Approach 2: Counting Set Bits)" << endl;
    } else {
        cout << "not a power of 2 (Approach 2: Counting Set Bits)" << endl;
    }

    cout << number << " is ";
    if (isPowerOfTwo_Approach3(number)) {
        cout << "a power of 2 (Approach 3: Log2)" << endl;
    } else {
        cout << "not a power of 2 (Approach 3: Log2)" << endl;
    }

    cout << number << " is ";
    if (isPowerOfTwo_Approach4(number)) {
        cout << "a power of 2 (Approach 4: Repeated Division)" << endl;
    } else {
        cout << "not a power of 2 (Approach 4: Repeated Division)" << endl;
    }

    cout << number << " is ";
    if (isPowerOfTwo_Approach5(number)) {
        cout << "a power of 2 (Approach 5: Bitset)" << endl;
    } else {
        cout << "not a power of 2 (Approach 5: Bitset)" << endl;
    }

    return 0;
}
